esphome/components/esp32_improv/esp32_improv_component.cpp

```diff
@@ -0,0 +1,263 @@
+      if (this->status_indicator_ != nullptr) {
+        if ((now % 200) < 100) {
+          this->status_indicator_->turn_on();
+        } else {
+          this->status_indicator_->turn_off();
+        }
+      }
+      if (wifi::global_wifi_component->is_connected()) {
+        wifi::global_wifi_component->save_wifi_sta(this->connecting_sta_.get_ssid(),
+                                                   this->connecting_sta_.get_password());
+        this->connecting_sta_ = {};
+        this->cancel_timeout("wifi-connect-timeout");
+        this->set_state_(improv::STATE_PROVISIONED);
+      }
+      break;
+    }
+    case improv::STATE_PROVISIONED: {
+      this->incoming_data_ = "";
+      if (this->status_indicator_ != nullptr)
+        this->status_indicator_->turn_on();
+      break;
+    }
+  }
+}
+
+bool ESP32ImprovComponent::check_identify_() {
+  uint32_t now = millis();
+
+  bool identify = this->identify_start_ != 0 && now - this->identify_start_ <= this->identify_duration_;
+
+  if (identify) {
+    uint32_t time = now % 1000;
+    if (time < 600 && time % 200 < 100) {
+      this->status_indicator_->turn_on();
+    } else {
+      this->status_indicator_->turn_off();
+    }
+  }
+  return identify;
+}
+
+void ESP32ImprovComponent::set_state_(improv::State state) {
+  ESP_LOGV(TAG, "Setting state: %d", state);
+  this->state_ = state;
+  if (this->status_->getData()[0] != state) {
+    ESP_LOGV(TAG, "States: %d -> %d", this->status_->getData()[0], state);
+    uint8_t data[1]{state};
+    this->status_->setValue(data, 1);
+    if (state != improv::STATE_STOPPED)
+      this->status_->notify();
+  }
+}
+
+void ESP32ImprovComponent::set_error_(improv::Error error) {
+  if (error != improv::ERROR_NONE)
+    ESP_LOGE(TAG, "Error: %d", error);
+  if (this->error_->getData()[0] != error) {
+    uint8_t data[1]{error};
+    this->error_->setValue(data, 1);
+    if (this->state_ != improv::STATE_STOPPED)
+      this->error_->notify();
+  }
+}
+
+void ESP32ImprovComponent::start() {
+  if (this->state_ != improv::STATE_STOPPED)
+    return;
+
+  ESP_LOGD(TAG, "Starting Improv service...");
+
+  this->service_->start();
+  BLEDevice::startAdvertising();
+  ESP_LOGD(TAG, "Service started!");
+
+  this->set_state_(improv::STATE_AWAITING_AUTHORIZATION);
+  this->error_->setValue({improv::ERROR_NONE});
+}
+
+void ESP32ImprovComponent::end() {
+  this->set_state_(improv::STATE_STOPPED);
+  this->service_->stop();
+  esp32_ble_server::global_ble_server->teardown();
+}
+
+float ESP32ImprovComponent::get_setup_priority() const {
+  // Before WiFi
+  return setup_priority::HARDWARE - 10.0f;
+}
+
+void ESP32ImprovComponent::dump_config() {
+  ESP_LOGCONFIG(TAG, "ESP32 Improv:");
+  LOG_BINARY_SENSOR("  ", "Activator", this->activator_);
+  ESP_LOGCONFIG(TAG, "  Status Indicator: '%s'", YESNO(this->status_indicator_ != nullptr));
+}
+
+void ESP32ImprovComponent::process_incoming_data_() {
+  uint8_t length = this->incoming_data_[1];
+
+  ESP_LOGD(
```

Comment:
```suggestion
  ESP_LOGV(
```

This feels more like verbose data (I know the naming of the log levels isn't great)